src/cudamatrix/cu-matrix-test.cc

```diff
@@ -1111,6 +1111,26 @@ template<typename Real> static void UnitTestCuMatrixAddMatMatElements() {
   KALDI_ASSERT(M.Sum() != 0.0);
 }
 
+template<typename Real> static void UnitTestCuMatrixAddMatMatDivMat() {
+  // M = a * b / c (by element; when c = 0, M = a)
+  MatrixIndexT dimM = 100 + Rand() % 255, dimN = 100 + Rand() % 255;
+  CuMatrix<Real> M(dimM, dimN), A(dimM, dimN), B(dimM, dimN), C(dimM, dimN);
+  CuMatrix<Real> ref(dimM, dimN);
+  M.SetRandn();
+  A.SetRandn();
+  B.SetRandn();
+  C.SetRandn();
+
+  M.AddMatMatDivMat(A,B,C);
+  ref.AddMatMatElements(1.0, A, B, 0.0);
```

Comment:
On Sun, Aug 28, 2016 at 2:53 AM, Daniel Povey <notifications@github.com>
wrote:

> In src/cudamatrix/cu-matrix-test.cc
> <https://github.com/kaldi-asr/kaldi/pull/1008#discussion_r76520755>:
>
> > @@ -1111,6 +1111,26 @@ template<typename Real> static void UnitTestCuMatrixAddMatMatElements() {
> >    KALDI_ASSERT(M.Sum() != 0.0);
> >  }
> >
> > +template<typename Real> static void UnitTestCuMatrixAddMatMatDivMat() {
> > +  // M = a * b / c (by element; when c = 0, M = a)
> > +  MatrixIndexT dimM = 100 + Rand() % 255, dimN = 100 + Rand() % 255;
> > +  CuMatrix<Real> M(dimM, dimN), A(dimM, dimN), B(dimM, dimN), C(dimM, dimN);
> > +  CuMatrix<Real> ref(dimM, dimN);
> > +  M.SetRandn();
> > +  A.SetRandn();
> > +  B.SetRandn();
> > +  C.SetRandn();
> > +
> > +  M.AddMatMatDivMat(A,B,C);
> > +  ref.AddMatMatElements(1.0, A, B, 0.0);
>
> There is a slight theoretical issue here, that the function
> AddMatMatDivMat has special behavior if an element of c is exactly 0. The
> prob. that this happens is very small, of course. we could just ignore it,
> but put a comment...
>
I think the comment about `c=0` is already there. Line 1115 and in
`cu-matrix.h` .

> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1008/files/7ff8e3153442b101bbd265b5579ea01d151e317c#r76520755>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADEVxHvkvlfHbTw0B3nBcT7jT9hHBi07ks5qkIeygaJpZM4Juv0r>
> .
>